packages/calypso-e2e/src/rest-api-client.ts

```diff
@@ -0,0 +1,157 @@
+import fetch, { BodyInit, HeadersInit, RequestInit } from 'node-fetch';
+import { SecretsManager } from './secrets';
+import type { AccountCredentials } from './data-helper';
+
+type EndpointVersions = '1' | '1.1' | '1.2' | '1.3';
+interface BearerTokenResponse {
+	success: string;
+	data: {
+		bearer_token: string;
+		token_links: string[];
+	};
+}
+interface RequestParams {
+	method: 'post' | 'get';
+	headers?: HeadersInit;
+	body?: BodyInit;
+}
+interface Site {
+	ID: string;
+	name: string;
+	description: string;
+	URL: string;
+	site_owner: string;
+}
+interface MySitesResponse {
+	sites: Site[];
+}
+
+const BEARER_TOKEN_URL = 'https://wordpress.com/wp-login.php?action=login-endpoint';
+const REST_API_BASE_URL = 'https://public-api.wordpress.com';
+
+/**
+ * Client to interact with the WordPress.com REST API.
+ */
+export class RestAPIClient {
+	readonly credentials: AccountCredentials;
+	private bearerToken: string | null;
+
+	/**
+	 * Constructs an instance of the API client.
+	 */
+	constructor( credentials: AccountCredentials ) {
+		this.credentials = credentials;
+		this.bearerToken = null;
+	}
+
+	/**
+	 * Returns the bearer token for the user.
+	 *
+	 * If the token has been previously obtained, this method returns the value.
+	 * Otherwise, an API call is made to obtain the bearer token.
+	 *
+	 * @returns {Promise<string>} String representing the bearer token.
+	 */
+	async getBearerToken(): Promise< void > {
+		if ( ! this.bearerToken === null ) {
+			return;
+		}
+
+		// Bearer Token generation is done via a URL query param.
+		const params = new URLSearchParams();
+		params.append( 'username', this.credentials.username );
+		params.append( 'password', this.credentials.password );
+		params.append( 'client_id', SecretsManager.secrets.calypsoOauthApplication.client_id );
+		params.append( 'client_secret', SecretsManager.secrets.calypsoOauthApplication.client_secret );
+		params.append( 'get_bearer_token', '1' );
+
+		// Request the bearer token for the user.
+		const response: BearerTokenResponse = await this.sendRequest( new URL( BEARER_TOKEN_URL ), {
+			method: 'post',
+			body: params,
+		} );
+
+		console.log( response );
+		this.bearerToken = response.data.bearer_token;
+	}
+
+	/* Request builder methods */
+
+	/**
+	 * Returns the appropriate authorization header.
+	 *
+	 * @returns {string} Authorization header in the requested scheme.
+	 * @throws {Error} If a scheme not yet implemented is requested.
+	 */
+	private getAuthorizationHeader( scheme: 'bearer' ): string {
```

Comment:
What do you think of making this async, and then having it call getBearerToken if the cached token is null?